source/common/router/upstream_request.cc

```diff
@@ -125,10 +125,17 @@ UpstreamRequest::UpstreamRequest(RouterFilterInterface& parent,
       parent_.callbacks()->connection(), parent_.callbacks()->streamId(),
       parent_.callbacks()->account(), true, parent_.callbacks()->decoderBufferLimit(),
       *parent_.cluster(), *this);
-  parent_.cluster()->createFilterChain(*filter_manager_);
+  bool created = parent_.cluster()->createFilterChain(*filter_manager_,
+                                                      /*only_create_if_configured=*/true);
+  if (!created) {
+    created = parent_.config().createFilterChain(*filter_manager_, false);
+  }
+  if (!created) {
+    created = parent_.cluster()->createFilterChain(*filter_manager_, false);
+  }
```

Comment:
yeah I'm confused at this. 
I thought this would create cluster-if-configured else create router no matter what.  
let's do e2e tests for the 4 config combinations

Reply:
Router doesn't have codec filter automatically applied; the cluster does. We could move that logic away from the cluster onto the router if you think that makes more sense?